_published/0014-test-results-handling.md

```diff
@@ -0,0 +1,257 @@
+    testResults, testSummary := parseTRX(file)
+    
+    for _, testResult := range testResults {
+    	testResult.ArtifactID = artifactID
+    }
+    testSummary.ArtifactID = artifactID
+    testSummary.BuildID = buildId
+    
+    m.Database.Create(&testResults)
+    m.Database.Create(&testSummary)
+}
+// new
+func storeArtifactInDB(file *File, buildID uint) (*Artifact) {
+    artifact := Artifact{
+    	Data: file.data, 
+    	Name: file.name, 
+    	FileName: file.fileName, 
+    	BuildID: buildID,
+    }
+    m.Database.Create(&artifact)
+    
+    return &artifact 
+}
+// new
+func parseMultipartFormData(c *gin.Context) []*File {
+    // ...
+}
+// new
+func parseTRX(file *File) []TestResult, TestResultSummary {
+    // ...
+}
+```
+
+```go
+// build.go
+// new, /build/{buildid}/test-result-summaries
+func (m BuildModule) getBuildTestResultSummariesHandler(c *gin.Context) {
+    buildId := ginutil.ParseParamUint(c, "buildid")
+
+    struct TestResultSummaries {
+    	Summaries []TestResultSummary `json:"summaries"`
+    	Count     uint                `json:"count"`
+    }
+    
+    testSummaries := TestResultSummaries{}
+    
+    m.Database.
+    	Where(&TestResultSummary{BuildID: buildId}).
+    	Find(&testSummaries.Summaries)
+    
+    if len(testSummaries.Summaries) > 0 {
+    	testResults.Count = len(testResults.Results)
+    } else {
+    	// dbnotfound error
+    	return
+    }
+    
+    // 200 with testSummaries
+}
+```
+
+</details>
+
+<details><summary>wharf-web</summary>
+
+wharf-web changes to use the new GET
+`/build/{buildid}/test-result-summaries` and GET `/build/{buildid}/artifact/{artifactid}/test-results`
+endpoints to retrieve the test result data instead of using the existing
+GET `/build/{buildid}/tests-results` endpoint.
+
+There would also be a way to view a build's test result details. [#17](https://github.com/iver-wharf/wharf-api/issues/17)
+</details>
+
+## Compatibility
+
+This breaks backward compatibility with projects using the
+GET `build/{buildid}/tests-results` endpoint, since it will get removed.
+
+## Alternative solutions
+
+Nothing comes to mind.
+
+## Future possibilities
+
+Nothing comes to mind.
+
+## Unresolved questions
+
+- I am having trouble thinking of how to test filling out the database
+  with data from the old test results. **What would be a good way to do this?**
+  
+- **Is an upload endpoint, separate from the one for other artifacts, for test
+  results necessary?**
+  I can see it being necessary if somebody wants to upload local test results
+  or something, but it doesn't feel like that would be required. ref. to: [Create a separate method to upload test results apart from artifacts](https://github.com/iver-wharf/wharf-api/issues/11)
+
+- As mentioned in [#17](https://github.com/iver-wharf/wharf-api/issues/17), we
+  are unlikely to have to store successful test details.
+  **Is there any foreseeable drawback to going that route?**
```

Comment:
Yes, you may want to see successful tests to track execution time over multiple runs (performance regression). I also think it should store which tests have been skipped so you can easily see if some test has been skipped dynamically by mistake.

Reply:
May want to make this opt-in. There's a lot of strings there (test names) we may not be interested in. The failed and skipped tests are a minority among the total number of tests, so they can stay in by default IMO.

It would be nice to be able to query the history of a certain test though, that's a very tempting and useful feature.